third_party/move/move-vm/runtime/src/storage/module_storage.rs

```diff
@@ -302,3 +398,51 @@ where
     )?;
     Ok(module.code().verified().clone())
 }
+
+/// Avoids the orphan rule to implement external [FunctionValueExtension] for any generic type that
+/// implements [ModuleStorage].
+pub struct FunctionExtensionAdapter<'a> {
+    pub(crate) module_storage: &'a dyn ModuleStorage,
+}
+
+pub trait AsFunctionValueExtension {
+    fn as_function_extension(&self) -> FunctionExtensionAdapter;
+}
+
+impl<T: ModuleStorage> AsFunctionValueExtension for T {
+    fn as_function_extension(&self) -> FunctionExtensionAdapter {
+        FunctionExtensionAdapter {
+            module_storage: self,
+        }
+    }
+}
+
+impl<'a> FunctionValueExtension for FunctionExtensionAdapter<'a> {
+    fn get_function_arg_tys(
+        &self,
+        module_id: &ModuleId,
+        function_name: &IdentStr,
+        ty_arg_tags: Vec<TypeTag>,
+    ) -> PartialVMResult<Vec<Type>> {
+        let ty_args = ty_arg_tags
```

Comment:
Maybe out of place comment, but we should start somewhere:

The ty_args, and ty names are used for different purposes. I'd have it:
`substitution_type_args` here, and in the lambda below `|ty| ....` -> `|type_to_substitute| ...`